File: backend/app/services/capabilities/tesla/climate.py
"""Tesla climate-related capabilities."""

from __future__ import annotations

from app.services.capabilities import register
from app.services.capabilities.base import (
    Capability,
    CapabilityCallContext,
    CapabilityResult,
    SafetyClass,
)


class SetClimateKeeperMode(Capability):
    """Set climate keeper mode: 0=off, 1=keep, 2=dog, 3=camp.

    Wraps `TeslaClient.set_climate_keeper_mode` (VCP-signed via
    tesla-http-proxy, keys on VIN).
    """

    @property
    def id(self) -> str:
        return "tesla.climate.set_keeper_mode"

    @property
    def safety_class(self) -> SafetyClass:
        return SafetyClass.WRITABLE

    @property
    def params_schema(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "mode": {
                    "type": "integer",
                    "enum": [0, 1, 2, 3],
                    "description": "0=off, 1=keep, 2=dog, 3=camp",
                }
            },
            "required": ["mode"],
        }

    async def invoke(
        self, ctx: CapabilityCallContext, params: dict
    ) -> CapabilityResult:
        mode = params.get("mode")
        if mode not in (0, 1, 2, 3):
            return CapabilityResult(success=False, error="mode must be 0..3")
        if not ctx.vin:
            return CapabilityResult(success=False, error="VIN required")
        await ctx.tesla_client.set_climate_keeper_mode(ctx.vin, mode)
        return CapabilityResult(success=True, data={"mode": mode})

    def expected_state(self, params: dict) -> dict:
        # After set_keeper_mode(N), the next telemetry frame should
        # show vehicle.climate.keeper_mode == N. Phase 9 confirmation.
        mode = params.get("mode")
        if mode not in (0, 1, 2, 3):
            return {}
        return {"vehicle.climate.keeper_mode": mode}


class Preheat(Capability):
    """Start HVAC (auto_conditioning_start) so the cabin is at temp
    on arrival. Used by the 出发前预热 automation."""

    @property
    def id(self) -> str:
        return "tesla.climate.preheat"

    @property
    def safety_class(self) -> SafetyClass:
        return SafetyClass.WRITABLE

    async def invoke(
        self, ctx: CapabilityCallContext, params: dict
    ) -> CapabilityResult:
        if not ctx.vin:
            return CapabilityResult(success=False, error="VIN required")
        await ctx.tesla_client.auto_conditioning_start(ctx.vin)
        return CapabilityResult(success=True)

    @property
    def dispatch_policy(self) -> str:
        # Preheat is time-sensitive: a "preheat at 7am" command that
        # finally lands at 9am because the car was asleep would
        # surprise the user. Drop instead of queue.
        return "drop_if_offline"


register(SetClimateKeeperMode())
register(Preheat())

File: apps/ios/Sources/TePlannerKit/Capabilities/Tesla/Climate.swift
import Foundation

/// `tesla.climate.set_keeper_mode` — sets climate keeper to off / keep
/// / dog / camp (0..3). Used by the camp-mode automation's "关闭"
/// action to dial the keeper back to off.
public struct SetClimateKeeperModeCapability: Capability {
    public let id = "tesla.climate.set_keeper_mode"
    public let safetyClass: SafetyClass = .writable

    public init() {}

    public func invoke(
        ctx: CapabilityContext,
        params: [String: JSONValue],
        api: APIServiceProtocol
    ) async -> CapabilityResult {
        guard let mode = params.int("mode"), (0...3).contains(mode) else {
            return CapabilityResult(success: false, error: "mode must be 0..3")
        }
        guard let vehicleId = ctx.vehicleId else {
            return CapabilityResult(success: false, error: "vehicleId required")
        }
        let result = await api.setClimateKeeperMode(vehicleId: vehicleId, mode: mode)
        switch result {
        case .success:
            return CapabilityResult(success: true)
        case .failure(let err):
            return CapabilityResult(success: false, error: err.localizedDescription)
        }
    }
}

/// `tesla.climate.preheat` — start HVAC (auto_conditioning_start) so
/// the cabin is at temperature on arrival.
public struct PreheatCapability: Capability {
    public let id = "tesla.climate.preheat"
    public let safetyClass: SafetyClass = .writable

    public init() {}

    public func invoke(
        ctx: CapabilityContext,
        params: [String: JSONValue],
        api: APIServiceProtocol
    ) async -> CapabilityResult {
        guard let vehicleId = ctx.vehicleId else {
            return CapabilityResult(success: false, error: "vehicleId required")
        }
        let result = await api.preheat(vehicleId: vehicleId)
        switch result {
        case .success:
            return CapabilityResult(success: true)
        case .failure(let err):
            return CapabilityResult(success: false, error: err.localizedDescription)
        }
    }
}
